### ISSUE TITLE
Code beautifier

### ISSUE THREAD

**Issue by benabraham** [mozilla/thimble.mozilla.org#1588]
Begginer coders using Thimble don't often have the skills to keep the code properly indented and generally well formatted. Such code is hard to read and debug and it makes learning much harder for them.

It would be great if [Brackets Beautify](https://github.com/brackets-beautify/brackets-beautify) plugin or directly [js-beautify](https://github.com/beautify-web/js-beautify) were integrated to Thimble. Even if activated via keyboard shortcut (so it wouldn't need GUI changes) it would mean great deal for my students.

**Comment by humphd** [comment]
I agree with you, and I like this idea a lot.  When we first shipped, I had the same intention, namely, to have beautifying be a feature we included.  I also added https://github.com/ahuth/brackets-paste-and-indent as a default extension, so that pasted code would match surrounding code indenting, and not leave new devs in a mess.

Adding this now is certainly possible.  I have a few concerns we'd have to sort out.  First, that we not confuse or frustrate users.  I'm not sure if having it run on Save will be good, since we autosave quite often and silently.  If beautifying on save is out, we need some kind of user initiated way to run it, whether new UI or a key combo as you suggest.  The problem with the former is that we risk cluttering the interface (our goal has always been to leave it as simple as possible, but even now it's gotten a bit complex).  The problem with the latter is that no one will ever discover it unless they know it exists, making it somewhat limited in its usefulness.  A third option might be to have it be part of the right-click context menu you get in the code editor.  Are there other options I'm not considering?

The second concern I have is that we not add size to the app, and make it slower to load.  Loading a bunch of JS libs that will only get used by a few people is not a great experience.  One way to deal with this would be to lazy-load the guts of this extension when they are first needed vs. on startup.  I've modified some other extensions and parts of Brackets to work like this.  We'd have to hack whatever extension we chose to do this late loading trick, which is a bit of work.

cc @Pomax, @gideonthomas, @flukeout

**Comment by benabraham** [comment]
You listed them probably all, except adding a button left of the undo and redo buttons, but that seems to be too much prominent space for such function.

Automatic is out of question, even if as option with default _off_ it would definitely confuse most users.

The best seems to add it to right-click context menu and a keyboard shortcut (if possible).

**Comment by gideonthomas** [comment]
I'm sorry for the delay in responding. I skimmed the contents of this thread but will definitely come back to add some thoughts about this soon.

Assigning myself so that I don't forget to follow up in this issue.
